jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/ResponseListeners.java

```diff
@@ -452,43 +454,35 @@ private void onDemandCallback()
 
         private void registerFailure(Throwable failure)
         {
-            while (true)
+            try (AutoLock ignored = lock.lock())
             {
-                long encoded = counters.get();
-                int failures = AtomicBiInteger.getHi(encoded) + 1;
-                int demands = AtomicBiInteger.getLo(encoded);
+                failures++;
+                if (failures == listeners.size())
+                    originalContentSource.fail(failure);
```

Comment:
Would not call fail() with lock held. Same for demand() below.

Also, if all have failed, we cannot fall through and also demand.

Reply:
Ok, let me change this.
There are other places where we call either fail() or demand() with a lock held. I'll change those too.